myia/opt/lib.py

```diff
@@ -1107,16 +1107,42 @@ def replace_Jinv_on_graph(resources, node, equiv):
     return ct
 
 
+def has_inner_j_on_function(node):
+    """Check if given node is a graph containing itself a J on a sub-graph."""
+    if isinstance(node, Graph):
+        seen = {node}
+        todo = [node.output]
+        while todo:
+            n = todo.pop()
+            if n.inputs:
+                for inp in n.inputs:
+                    if inp.is_apply(P.J) and inp.inputs[1].is_constant(Graph):
+                        return True
+                    else:
+                        todo.append(inp)
+            elif n.is_constant_graph() and n.value not in seen:
+                seen.add(n.value)
+                todo.append(n.value.output)
```

Comment:
Using the manager, I think this would be equivalent to something like:

```python
for graph in node.graphs_reachable:
    if any(x.is_apply(P.J) and x.inputs[1].is_constant(Graph) for x in graph.nodes):
        return True
return False
```

Or possibly `node.scope` instead of `node.graphs_reachable`. The former only looks into the closures, so it would be more economical.

Reply:
Hi @breuleux ! `scope` and `graphs_reachable` seems to be members of `GraphManager` class only, not `ANFNode` class.

Also, I need to check not only closures, but also any apply node in the sub-graph of given node (ie. not in the entire graph that includes the given node), so I'm not sure that visiting all reachable graphs or graphs in scope will pass through where I need.

Reply:
They should be members of Graph if I'm not mistaken, and node is an instance of Graph in that branch. Either way, fair enough, it's true that doesn't do exactly the same thing.